### ISSUE TITLE
Protobuf UserMessages leaking memory

### ISSUE THREAD

**Issue by neko-pm** [alliedmodders/sourcemod#1141]
After a fair bit of testing I've narrowed down the cause of a memory leak I've had. The following code causes the memory usage to climb rather quickly, causing the server to eventually shutdown.
```SP
Protobuf hMessage = view_as<Protobuf>(StartMessageEx(g_TextMsg, players, count, USERMSG_RELIABLE|USERMSG_BLOCKHOOKS));

if(hMessage)
{
	hMessage.SetInt("msg_dst", 4);
	hMessage.AddString("params", "#SFUI_ContractKillStart");
	hMessage.AddString("params", message);
	hMessage.AddString("params", NULL_STRING);
	hMessage.AddString("params", NULL_STRING);
	hMessage.AddString("params", NULL_STRING);
}

EndMessage();
```
The speed that the memory usage climbed at seemed to be correlated to the size of the `message` string.

I have tried to figure out what the underlying cause is, but I am still unsure. However, what I did do was create an extension to test with. Since using the following code, in place of the above sourcepawn implementation, I have not observed the same fast exhaustion of memory.
```CPP
bool FixedTextMsg(cell_t* players, int playersNum, const char* message)
{
	static auto CCSUsrMsg_TextMsgFactory = &CCSUsrMsg_TextMsg::default_instance();
	CCSUsrMsg_TextMsg* pMsg = CCSUsrMsg_TextMsgFactory->New();

	if(pMsg == nullptr)
		return false;

	CellRecipientFilter CellRecFilter;
	CellRecFilter.Initialize(players, playersNum);
	CellRecFilter.SetToReliable(true);

	pMsg->set_msg_dst(4);
	pMsg->add_params("#SFUI_ContractKillStart");
	pMsg->add_params(message);
	pMsg->add_params("");
	pMsg->add_params("");
	pMsg->add_params("");

	ENGINE_CALL(SendUserMessage)(static_cast<IRecipientFilter &>(CellRecFilter), CS_UM_TextMsg, *pMsg);
	
	delete pMsg;

	return true;
}
```

Additional Information:
```
OS: Windows

SourceMod Version Information:
  SourceMod Version: 1.10.0.6455
  SourcePawn Engine: 1.10.0.6455, jit-x86 (build 1.10.0.6455)
  SourcePawn API: v1 = 5, v2 = 12
  Compiled on: Nov 18 2019 19:39:08
  Built from: https://github.com/alliedmodders/sourcemod/commit/e135eff1
  Build ID: 6455:e135eff1

Metamod:Source version 1.10.7-dev
Built from: https://github.com/alliedmodders/metamod-source/commit/6c8495f
Build ID: 971:6c8495f
Loaded As: Valve Server Plugin
Compiled on: Sep  2 2019
Plugin interface version: 15:14
SourceHook version: 5:5
```
I'm happy to provide additional information if I've missed anything.

**Comment by asherkin** [comment]
I can't see any meaningful difference between the extension code you wrote and the SM implementation that is being exercised with the SP code you've replaced.

I wrote this simple test plugin and played with bots for a few rounds and observed no increase in memory usage while profiling:
```sourcepawn
#include <sourcemod>

#pragma semicolon 1
#pragma newdecls required

UserMsg g_TextMsg = INVALID_MESSAGE_ID;

public void OnMapStart()
{
	g_TextMsg = GetUserMessageId("TextMsg");
}

public void OnClientPutInServer(int client)
{
	int userId = GetClientUserId(client);
	CreateTimer(0.1, OnTimer, userId, TIMER_REPEAT);
}

Action OnTimer(Handle timer, int userId)
{
	int client = GetClientOfUserId(userId);
	if (client <= 0) {
		return Plugin_Stop;
	}
	
	int players[1];
	players[0] = client;
	int count = 1;
	
	Protobuf hMessage = view_as<Protobuf>(StartMessageEx(g_TextMsg, players, count, USERMSG_RELIABLE | USERMSG_BLOCKHOOKS));
	if(!hMessage) {
		return Plugin_Stop;
	}
	
	char message[4096];
	FormatEx(message, sizeof(message), "%4096s", "Hello, World!");
	
	hMessage.SetInt("msg_dst", 4);
	hMessage.AddString("params", "#SFUI_ContractKillStart");
	hMessage.AddString("params", message);
	hMessage.AddString("params", NULL_STRING);
	hMessage.AddString("params", NULL_STRING);
	hMessage.AddString("params", NULL_STRING);
	EndMessage();
	
	return Plugin_Handled;
}
```

I'm afraid that I think we're going to need more information to go on here - quite a lot of people have been seeing issues with memory usage in CS:GO recently which does coincide with the TextMsg workarounds but this is the first thing pointing towards the leak being in SM.

**Comment by neko-pm** [comment]
> I can't see any meaningful difference between the extension code you wrote and the SM implementation that is being exercised with the SP code you've replaced.

That's what confused me, I wrote that extension code as a starting point to get a better understanding of what was happening, but it ended up not exhibiting the same issue. 
![unknown](https://user-images.githubusercontent.com/20050672/70869987-2f6d6a80-1f86-11ea-830f-b29d9f29739d.png)
As you can see (ignoring the slow creep still present), after making a single change, which was removing the sourcepawn implementation from the initial post, the fast usage of memory stopped. Adding it back and watching the real time usage readout, I could see it climbing again like it did before.

For some more context, I'm sending messages every tick, to every player and their spectators. In my testing, I removed sections of the plugin at a time, which is how I came to the conclusion it was the protobuf messages, rather than other parts of the code.

The only main difference I really see between the code you've used and mine, is that mine is executed from within a `RequestFrameCallback` with parameters in a DataPack, which is called from a `MsgHook` callback. I also made sure that the issue didn't stem from the DataPack being misused.

**Comment by asherkin** [comment]
> The only main difference I really see between the code you've used and mine, is that mine is executed from within a `RequestFrameCallback` with parameters in a DataPack, which is called from a `MsgHook` callback. I also made sure that the issue didn't stem from the DataPack being misused.

Could the problem be in the hooking side rather than the sending side? Does your extension version still utilise the SM/SP hooks?

**Comment by neko-pm** [comment]
The extension code is called from the `MsgHook` callback, though a native, as it seemed like the limitation of waiting a frame was for compatibility reason in csgo's case. This is in place of using `RequestFrame` and then the sourcepawn code.

**Comment by CrazyHackGUT** [comment]
I'm not sure, but looks like SM doesn't remove fully `protobuf::Message` after message sending.
https://github.com/alliedmodders/sourcemod/blob/8031e42bdae3046781b5d52a7a90d59e788105e6/core/UserMessagePBHelpers.h#L98-L109
There should be a deletion of `msg` (`protobuf::Message`) too.

**Comment by asherkin** [comment]
> I'm not sure, but looks like SM doesn't remove fully `protobuf::Message` after message sending.

The `protobuf::Message` ptr gets preserved by the caller (either the UserMessage subsystem, or the hook implementations) and free'd appropriately. There is one case that isn't quite as clear cut and that is reading internal message fields (the protobuf docs are unclear on ownership), but it looks correct and seems right in testing (i.e. I added code to free it and the game started crashing).

I've done a load more testing with freeing different things and profiling memory usage using a whole heap of plugins doing different things and I still can't reproduce this leak, so I'm not sure what's going on here. It definitely exists in the wild, it definitely started at the same time as the hint message workarounds, and it definitely seems to be avoided by not implementing them in SourcePawn.

Has anyone seen it on Windows, or only Linux?

**Comment by neko-pm** [comment]
> Has anyone seen it on Windows, or only Linux?

I was experiencing the issue on Windows, didn't have a chance to test in on Linux.

**Comment by hlstriker** [comment]
I have [this plugin](https://github.com/Franc1sco/FixHintColorMessages/blob/master/FixHintColorMessages.sp) on my servers and noticed it was causing massive memory leaks and eventually crashing the server. The plugin is similar to what you are describing and since you mentioned you used data packs I have a feeling your memory leak is caused by [this bug](https://github.com/alliedmodders/sourcemod/pull/1251).

**Comment by asherkin** [comment]
I'm going to close this issue as it seems like the Datapack leak was the cause, and the fix for that has now been merged to 1.10-dev. Please report back if it doesn't seem to have solved the issue!
